sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py

```diff
@@ -663,17 +663,21 @@ def _try_resolve_compute_for_node(cls, node: BaseNode, _: str, resolver):
 
     @classmethod
     def _divide_nodes_to_resolve_into_layers(cls, component: PipelineComponent, extra_operations: List[Callable]):
-        """Traverse the pipeline component and divide nodes to resolve into layers.
+        """Traverse the pipeline component and divide nodes to resolve into layers. Note that all leaf nodes will be
+        put in the last layer.
         For example, for below pipeline component, assuming that all nodes need to be resolved:
           A
          /|\
         B C D
         | |
         E F
+        |
+        G
         return value will be:
         [
-          [("B", B), ("C", C), ("D", D)],
-          [("E", E), ("F", F)],
+          [("B", B), ("C", C)],
```

Comment:
Why do we need to spilt the nodes into layers before resolving them?
Seems we are not splitting them according to depth now

Reply:
1. In DPv2, we use iteration instead of recursion to resolve subgraphs to get a better control of the process. For example, we can avoid duplicate processing for a component instance referred in multiple subgraphs in an easier way.
2. when using iteration, we should manage the order to resolve nodes manually - a pipeline component can't be resolved before all its child nodes have been resolved. We use this split function to simplify the management - high layer depends on low layers.
3. Before this PR, we do split nodes according to depth. But 1 problem is node resolution performance is not evenly distributed. Leaf node resolution may involve **local IO**, **artifact downloading** and **file uploading**, so usually much slower than pipeline nodes. We move all leaf nodes to the last layer to do slow operations concurrently. Below is a more detailed example:
```
              A
      /       |       \
B(1s)  C (1s)  D (10s)
   |          |
E(1s)   F(10s)
  |
G(10s)
```
If we simply process nodes according to depth, it will take **30s** to resolve all nodes; if we move all leaf nodes to the last layer, it will take **12s** to resolve all nodes.